monai/networks/nets/vista3d.py

```diff
@@ -0,0 +1,951 @@
+
+    def update_point_to_patch(self, patch_coords, point_coords, point_labels):
+        """ Update point_coords with respect to patch coords.
+        If point is outside of the patch, remove the coordinates and set label to -1
+        """
+        patch_ends = [
+            patch_coords[-3].stop,
+            patch_coords[-2].stop,
+            patch_coords[-1].stop,
+        ]
+        patch_starts = [
+            patch_coords[-3].start,
+            patch_coords[-2].start,
+            patch_coords[-1].start,
+        ]
+        # update point coords
+        patch_starts = (
+            torch.tensor(patch_starts, device=point_coords.device)
+            .unsqueeze(0)
+            .unsqueeze(0)
+        )
+        patch_ends = (
+            torch.tensor(patch_ends, device=point_coords.device)
+            .unsqueeze(0)
+            .unsqueeze(0)
+        )
+        # [1 N 1]
+        indices = torch.logical_and(
+            ((point_coords - patch_starts) > 0).all(2),
+            ((patch_ends - point_coords) > 0).all(2),
+        )
+        # check if it's within patch coords
+        point_coords = point_coords.clone() - patch_starts
+        point_labels = point_labels.clone()
+        if indices.any():
+            point_labels[~indices] = -1
+            point_coords[~indices] = 0
+            # also remove padded points, mainly used for inference.
+            not_pad_indices = (point_labels != -1).any(0)
+            point_coords = point_coords[:, not_pad_indices]
+            point_labels = point_labels[:, not_pad_indices]
+        else:
+            point_coords = None
+            point_labels = None
+        return point_coords, point_labels
+
+    def connected_components_combine(
+        self, logits, point_logits, point_coords, point_labels, mapping_index, thred=0.5
+    ):
+        """Combine auto results with point click response, or combine previous mask with point click response.
+        For mapping_index with point clicks, NaN values in logits will be replaced with point_logits. Meanwhile, the added/removed
+        region in point clicks must be updated by the lcc function. Notice, if a positive point is within logits/prev_mask, the components containing the positive point
+        will be added.
+        """
+        logits = (
+            logits.as_tensor() if isinstance(logits, monai.data.MetaTensor) else logits
+        )
+        _logits = logits[mapping_index]
+        inside = []
+        for i in range(_logits.shape[0]):
+            inside.append(
+                np.any(
+                    [
+                        _logits[
+                            i,
+                            0,
+                            round(p[0].item()),
+                            round(p[1].item()),
+                            round(p[2].item()),
+                        ].item()
+                        > 0
+                        for p in point_coords[i]
+                    ]
+                )
+            )
```

Comment:
I think `point_coords` should have a rounded copy made if it's not too large so that `round` isn't called in the comprehension, this would reduce how large this whole expression is. If `point_coords` is too long I think this should be broken up into multiple statements for better readability.